packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,360 @@
+      // add tag to connection describing discovery mechanism
+      this.libp2pComponents.connectionManager
+        .getConnections(peerId)
+        .forEach((conn) => conn.tags.push(...tags));
+
+      this.dialAttemptsForPeer.delete(peerId.toString());
+    } catch (error) {
+      const dialAttempt = this.dialAttemptsForPeer.get(peerId.toString()) ?? 1;
+      this.dialAttemptsForPeer.set(peerId.toString(), dialAttempt + 1);
+
+      //remove peer from peerStore if dial fails after max attempts
+      if (dialAttempt > maxDialAttemptsForPeer) {
+        try {
+          log(
+            "Deleting undialable peer " + peerId.toString() + "from peerStore"
+          );
+          await this.libp2pComponents.peerStore.delete(peerId);
+          return;
+        } catch (error) {
+          throw (
+            "Error deleting undialable peer" +
+            peerId.toString() +
+            " from peerStore"
+          );
+        }
+      }
+
+      log(
+        `Failed to dial ${peerId.toString()} - reattempting (${dialAttempt})`
+      );
+
+      this.dialPeer(peerId);
+    }
+  }
+
+  private startPeerDiscoveryListener(): void {
+    const onDiscovery = async (evt: CustomEvent<PeerInfo>): Promise<void> => {
+      const { id: peerId } = evt.detail;
+
+      if (!(await this.isPeerDialable(peerId))) return;
+
+      this.dialPeer(peerId).catch((err) =>
+        log(`Error dialing peer ${peerId.toString()} : ${err}`)
+      );
+    };
+
+    this.libp2pComponents.peerStore.addEventListener("peer", onDiscovery);
+  }
+
+  private startPeerConnectionListener(
+    keepAliveOptions: KeepAliveOptions,
+    relay?: IRelay
+  ): void {
+    const onConnect = (evt: CustomEvent<Connection>): void => {
+      {
+        this.startKeepAlive(
+          evt.detail.remotePeer,
+          this.libp2pComponents.ping.bind(this),
+          keepAliveOptions,
+          relay
+        );
+      }
+    };
+
+    this.libp2pComponents.connectionManager.addEventListener(
+      "peer:connect",
+      onConnect
+    );
+  }
+
+  private startPeerDisconnectionListener(): void {
+    // TODO: ensure that these following issues are updated and confirmed
+    /**
+     * NOTE: Event is not being emitted on closing nor losing a connection.
+     * @see https://github.com/libp2p/js-libp2p/issues/939
+     * @see https://github.com/status-im/js-waku/issues/252
+     *
+     * >This event will be triggered anytime we are disconnected from another peer,
+     * >regardless of the circumstances of that disconnection.
+     * >If we happen to have multiple connections to a peer,
+     * >this event will **only** be triggered when the last connection is closed.
+     * @see https://github.com/libp2p/js-libp2p/blob/bad9e8c0ff58d60a78314077720c82ae331cc55b/doc/API.md?plain=1#L2100
+     */
+    this.libp2pComponents.connectionManager.addEventListener(
+      "peer:disconnect",
+      (evt) => {
+        this.stopKeepAlive(evt.detail.remotePeer);
+      }
+    );
+  }
```

Comment:
Any reason why having this keep alive logic in the connection manager? I'd move it to the keep alive manager.

Reply:
Well, the the keep alive logic applies to a given libp2p instance and starts for a peer when it connects and stop for a peer when it disconnects. I would move this logic in the keep alive manager, whose role is to keep alive connection to peers.